File: backend/src/lua/tokenBucket.lua
-- KEYS[1] = bucket key

-- ARGV[1] = current timestamp (ms)
-- ARGV[2] = capacity
-- ARGV[3] = refill rate (tokens/sec)
-- ARGV[4] = bucket ttl (seconds)

local key = KEYS[1]

local now = tonumber(ARGV[1])

local capacity = tonumber(ARGV[2])

local refillRate = tonumber(ARGV[3])

local ttl = tonumber(ARGV[4])


-- Read bucket

local bucket = redis.call("HMGET", key, "tokens", "lastRefill")

local tokens = tonumber(bucket[1])

local lastRefill = tonumber(bucket[2])


-- First request

if tokens == nil then

    tokens = capacity

    lastRefill = now

end



local elapsed = (now - lastRefill) / 1000


-- Refill

tokens = math.min(

    capacity,

    tokens + elapsed * refillRate

)


-- No tokens

if tokens < 1 then

    local retryAfter = math.ceil((1 - tokens) / refillRate)

    return {

        0,

        tokens,

        retryAfter

    }

end


-- Consume token

tokens = tokens - 1


-- Save bucket

redis.call(

    "HMSET",

    key,

    "tokens",

    tokens,

    "lastRefill",

    now

)

redis.call("EXPIRE", key, ttl)


return {

    1,

    tokens,

    0

}

File: backend/src/services/rateLimiter/tokenBucket.js
const {client} = require("../../config/redis");
const { tokenBucketScript } = require("../../utils/luaScripts");
const rateLimitConfig = require("../../config/rateLimitConfig");

const tokenBucket = async (req, res, next) => {
    try {
        const userPlan = req.user.plan || "free";

        if (userPlan === "admin") {
            return next();
        }

        // Current config (derived refill rate)
        const {
            capacity: CAPACITY,
            refillRate,
            ttl,
        } = rateLimitConfig.plans[userPlan].tokenBucket;

        // // Tokens generated per second
        // const refillRate = CAPACITY / WINDOW;

        const key = `token_bucket:${req.user.id}`;

        const now = Date.now();

        const result = await client.eval(tokenBucketScript, {
            keys: [key],
            arguments: [
                String(now),
                String(CAPACITY),
                String(refillRate),
                String(ttl),
            ],
        });

        const allowed = Boolean(Number(result[0]));
        const tokensLeft = Number(result[1]);
        const retryAfter = Number(result[2]);

        res.set({
            "X-RateLimit-Limit": CAPACITY,
            "X-RateLimit-Remaining": Math.floor(tokensLeft),
            "Retry-After": retryAfter,
        });

        if (!allowed) {
            return res.status(429).json({
                success: false,
                message: "Too Many Requests",
                retryAfter,
            });
        }

        next();

    } catch (error) {

        console.error(error);

        return res.status(500).json({
            success: false,
            message: "Token Bucket Error",
        });

    }
};

module.exports = tokenBucket;